python/dgl/convert.py

```diff
@@ -0,0 +1,722 @@
+        metagraph, [rgrh._graph for rgrh in rel_graphs])
+    return DGLHeteroGraph(hgidx, ntypes, etypes)
+
+def to_hetero(G, ntypes, etypes, ntype_field=NTYPE, etype_field=ETYPE):
+    """Convert the given graph to a heterogeneous graph.
+
+    The input graph should have only one type of nodes and edges. Each node and edge
+    stores an integer feature (under ``ntype_field`` and ``etype_field``), representing
+    the type id, which which can be used to retrieve the type names stored
+    in the given ``ntypes`` and ``etypes`` arguments.
+
+    Examples
+    --------
+    TBD
+
+    Parameters
+    ----------
+    G : DGLHeteroGraph
+        Input homogenous graph.
+    ntypes : list of str
+        The node type names.
+    etypes : list of str
+        The edge type names.
+    ntype_field : str, optional
+        The feature field used to store node type. (Default: dgl.NTYPE)
+    etype_field : str, optional
+        The feature field used to store edge type. (Default: dgl.ETYPE)
+
+    Returns
+    -------
+    DGLHeteroGraph
+        A heterograph.
+        The parent node and edge ID are stored in the column dgl.NID and dgl.EID
+        respectively for all node/edge types.
+
+    Notes
+    -----
+    The returned node and edge types may not necessarily be in the same order as
+    ``ntypes`` and ``etypes``.  And edge types may be duplicated if the source
+    and destination types differ.
+
+    The node IDs of a single type in the returned heterogeneous graph is ordered
+    the same as the nodes with the same ``ntype_field`` feature.  Edge IDs of
+    a single type is similar.
+    """
+    # TODO(minjie): use hasattr to support DGLGraph input; should be fixed once
+    #  DGLGraph is merged with DGLHeteroGraph
+    if (hasattr(G, 'ntypes') and len(G.ntypes) > 1
+            or hasattr(G, 'etypes') and len(G.etypes) > 1):
+        raise DGLError('The input graph should be homogenous and have only one '
+                       ' type of nodes and edges.')
+
+    ntype_ids = F.asnumpy(G.ndata[ntype_field])
+    etype_ids = F.asnumpy(G.edata[etype_field])
+
+    num_ntypes = np.max(ntype_ids) + 1
+    num_etypes = np.max(etype_ids) + 1
+
+    def _group_by(ids, val_range):
+        return [np.where(ids == v)[0] for v in range(val_range)]
+
+    node_groups = _group_by(ntype_ids, num_ntypes)
+    edge_groups = _group_by(etype_ids, num_etypes)
+
+    src, dst = G.all_edges(order='eid')
+    src = F.asnumpy(src)
+    dst = F.asnumpy(dst)
+
+    # relabel nodes and edges
+    def _make_invmap(ids, invmap):
+        invmap[ids] = np.arange(len(ids))
+
+    node_invmap = np.zeros((G.number_of_nodes(),), dtype=src.dtype) - 1
+    edge_invmap = np.zeros((G.number_of_edges(),), dtype=src.dtype) - 1
+    for grp in node_groups:
+        _make_invmap(grp, node_invmap)
+    for grp in edge_groups:
+        _make_invmap(grp, edge_invmap)
+
+    # infer metagraph
+    canonical_etids = []
+    for etid in range(num_etypes):
```

Comment:
So are we not allowing the users to have the same edge type ID if the source-destination type pairs differ?  I think we should allow this if we allow such edge types to have the same name.

Reply:
Same name is allowed. For example, there could be two `"watches"` in the `etypes` argument. However, I feel the IDs should be different. Otherwise, `len(G.etypes)` might not be equal to `len(G.canonical_etypes)`, which is quite strange to me.

Reply:
No, I mean we need to detect the same edge IDs with different source-destination types, and translate them to different canonical etypes, just like what we did for edge type names.

Reply:
On the other hand, I don't think we need to maintain the order of edge types in the heterograph to be the same as that in  the `etypes` argument, because after the users called `to_hetero()`, we can safely assume that from then on the users would only access the features and do MP with the edge type names instead of IDs.

Reply:
For the user provided edge type ID, the question is whether the edge ID is similar to edge type name that allows collision. Current implementation is NO. That is to say the provided edge ID one-one corresponds to the canonical edge type. Your proposal is YES. I think both proposals are OK. Hence the question is how easy (efficient) will the implementation be? You could see that the current implementation avoids looping over all the nodes/edges in python and only loops over node/edge types which is a nice property I want to keep. If edge ID can have collision, is this still achievable? I feel inferring the metagraph will be quite painful (i.e, need to compute some complicated unique operation).

Reply:
For the edge type order, I think keeping the order makes it easier to jump from IDs to actual names. If we could keep it without too much of an effort, let's try to have that.

Reply:
> Hence the question is how easy (efficient) will the implementation be?

I don't think sacrificing usability for efficiency is a good rationale in particular for this case.

Essentially, we would be leaving the de-duplication of edge types to functions at higher levels.  For instance, converting from NetworkX/Neptune/Neo4j/... graphs with edge types as strings would lead us into a similar problem.  How would we handle those cases?  If we were to figure out the different canonical etypes for the same etype in those routines, then why not handle them in `to_hetero()`, so that the conversion from NetworkX/Neptune/etc only involves (1) transform the graph into a homograph with features, and (2) call `to_hetero()`?

Reply:
Ok. I'm convinced that such functionality might be inevitable. However, we should avoid looping over graphs in python as it will be too slow for even moderate size graph. At the moment, I cannot think of a good way to implement this using numpy.

For the discussion of user experience, I want to use an example to check we are on the same page. Suppose a graph has two relations: `user-watches-TV` `user-watches-movie`.
(1) What `G.etypes` should be?
(2) What `G.canonical_etypes` should be?
(3) What is the `etype` argument to `dgl.to_hetero`?
(4) Given the above `etype`, if there are two edges (one for each type), what the type ID should be?

For the current implementation,
(1) `['watches', 'watches']`
(2) `[('user', 'watches', 'TV'), ('user', 'watches', 'movie')]`
(3) `['watches', 'watches']`
(4) `[0, 1]`

Reply:
Let's say we have the following heterograph:

```
TV #0 <--(watches)-- User #0 --(watches)--> Movie #0 
```

Then I would expect

```python
g = dgl.graph([(1, 0), (1, 2)])
g.ndata[NTYPE] = torch.LongTensor([0, 1, 2])
# Here I would expect 0 to be just an index to the `etypes` argument.
g.edata[ETYPE] = torch.LongTensor([0, 0])
hg = dgl.to_hetero(g, ntypes=['TV', 'user', 'movie'], etypes=['watches'])
assert set(hg.ntypes) == {'user', 'TV', 'movie'}
assert set(hg.canonical_etypes) == {
        ('user', 'watches', 'TV'), ('user', 'watches', 'movie')}
```

Reply:
For the purpose of efficiency, I think we will eventually need to migrate this into C++.  We can do that in 0.5.

Reply:
It's better be done this release. This is a very common utility and I'm afraid that the efficiency may make this API not usable.